hrc_speech_prediction/models.py

```diff
@@ -90,12 +150,183 @@ def fit(self, X_context, X_speech, labels):
         self._q = np.power(probas, self.beta).sum(axis=0) + 1.e-8
         return self
 
     def _predict_proba(self, X_context, X_speech):
-        probas = super(PragmaticModel, self)._predict_proba(X_context, X_speech)
-        l_probas = (np.power(probas, self.beta) *
-                    self._p[None, :] / self._q[None, :])
+        probas = super(PragmaticModel, self)._predict_proba(
+            X_context, X_speech)
+        l_probas = (
+            np.power(probas, self.beta) * self._p[None, :] / self._q[None, :])
         return normalize(l_probas, axis=1)
+
+
+class CombinedModel(object):
+    def __init__(self,
+                 vectorizer,
+                 model_generator,
+                 actions,
+                 speech_eps=0.15,
+                 context_eps=0.15):
+
+        self._speech_eps = speech_eps
+        self._context_eps = context_eps
+
+        self.actions = actions
+
+        self.model_generator = model_generator
+        self._vectorizer = vectorizer
+
+        self.context_model = context_model.ContextTreeModel(
+            self.actions, eps=context_eps)
+        self.speech_model = None
+
+        self._X_context = np.ones((1, len(self.actions)), dtype='bool')
+
+    def fit(self, ctxt, speech, actions):
+        self.context_model.fit(ctxt, actions)
+        self.speech_model = self.model_generator(
+            self.actions, features="speech").fit(
+                self._X_context, speech, actions, sample_weight=None)
+
+    def partial_fit(self, ctxt, speech, action):
+
+        if isinstance(speech, str):
+            x_u = self._vectorizer.transform([speech])
+        else:
+            x_u = speech  # Then the input is an numpy array already
+
+        self.context_model.fit(ctxt, action)
+
+        if self.speech_model:
+            self.speech_model.partial_fit(self._X_context, x_u, action,
+                                          self.actions)
+        else:
+            self.speech_model = self.model_generator(
+                self.actions, features="speech").partial_fit(
+                    self._X_context, x_u, action, classes=self.actions)
+
+    def predict(self,
+                cntxt,
+                utter=None,
+                model="both",
+                exclude=None,
+                plot=False,
+                return_probs=False):
+
+        if model == "context" is not None:
+            probs = self.get_context_probs(cntxt)
+        elif model == "speech" and utter is not None:
+            probs = self.get_speech_probs(utter)
+        elif model == "both" and utter is not None:
+            context_probs = self.get_context_probs(cntxt)
+            speech_probs = self.get_speech_probs(utter)
+        else:
+            raise "Error, bad inputs to predict()!"
+
+        probs = np.multiply(context_probs, speech_probs)
+
+        action = self.get_probable_action(probs, exclude)
+        # self.curr = self.curr._get_next_node(action)
+
+        if plot and model == "both":
```

Comment:
Honestly, you don't gain much to have the plot here within the predict method. You could put that in another method instead and just pass it the probabilities.

Also, try avoiding calls to global matplotlib functions within a plot function. You don't know what code is calling this method and whether the plot is one among others (in which case you don't want the `gcf().clear()`), or needs to be shown or saved to a file (in which case you don't want the `show()`. All these belong to your controller code.